rfcs/0000-light-rail-trip-model.md

```diff
@@ -0,0 +1,211 @@
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+The primary drawback of doing this is if we decide that the benefits outlined in the [Motivation](#motivation) section are not worth the effort. There are also potential downsides to having Glides become the source of truth for trip assignment data; some of those are discussed in the [Rationale and alternatives](#rationale-and-alternatives) section below.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+## Glides versus RTR as the source of truth for assignments
+
+In deciding which system would be the source of truth for assignments between trains and trips, I had to weight the pros and cons of both RTR and Glides taking that logic on, ultimately landing on having it live in Glides. The following factors influenced my decision:
+
+1. Assigning trips to trains is a fundamentally stateful exercise that depends on a trains progress on its previous trip, location, trainsheet data about upcoming trips, and other factors. RTR has some stateful logic, but its state storage and management system is not very advanced. Glides has a relational database that is already updated with information about train movements and trips as they happen.
+1. This mirrors the situation on heavy rail where the OCS performs realtime trip assignments and RTR uses those as its main source of input for heavy rail trip assignments.
+
+The main downsides of putting this logic in Glides are:
+
+1. As Glides will be more directly responsible for data that flows more directly into passenger-facing systems, it is possible that the Glides team will have to respond to more feedback about issues related to why Green Line trains were assigned to incorrect trips. However, this is likely going to happen anyway with trainsheet-based predictions in RTR given that that already relies on input from Glides.
+1. Glides does not currently have a way of getting location data on yard pull-out / pull-back trains from RTR, so we will need to set up a pipeline for that in order for Glides to implement the full functionality. However, we already need to do this to enable other functionality in Glides, so it is not introducing net new work.
+1. Glides will need to generate automatically-created trips for trains in service that aren't matched to any trip on a trainsheet, logic that is currently implemented in RTR. However, this is likely desirable anyway because Glides will need a place to associate things like trip notes on unmatched trains in the future.
+1. Glides may not be aware of some disruptions in its schedule data that are modeled at the `gtfs_creator` level, meaning that we will still need RTR to create ADDED trips for any scheduled trips that it doesn't recognize. With the current logic, RTR will at least attempt to match to trips in the disrupted GTFS. However, given that these trips aren't guaranteed to match actual operational practice, it's unclear how much value there is in matching to them anyway.
```

Comment:
I think I may be missing something here. Glides train - trip assignment events seemed to describe glides producing GTFS trip ids for trips it can match, and added trips ids for those it can't. Under what circumstances is RTR still responsible for generating it's own trip ids?

Reply:
This is the relevant section from the "Glides train - trip assignment events" section above:

> If RTR receives a vehicle_trip_assignment event for a trip ID that it does not recognize (via either the GTFS schedule or via a trip add trips_updated event), it SHOULD map that to a unique added trip ID of its own and maintain that mapping in the GTFS-rt feed for the duration of the trip (that is, the unrecognized trip ID from Glides should not suddenly become mapped to a different randomly-generated added trip)

So in this case, RTR generates an ID but it should be noted that it's a bit different from the present case, as RTR is not responsible for the full act of assigning a trip, just the ID. The reason I want to avoid simply passing a trip ID from Glides into the public GTFS-rt feed if it doesn't match a trip in the currently active GTFS schedule is that it may be the ID of a trip that is in the GTFS file bot not active on the current service date due to trip ID changes around shuttle activations, and I don't want downstream consumers to potentially be confused by that.

Reply:
Got it, I think that makes sense. We should still guard against in, but in practice how likely is this? Picking on B again, it's currently closed past Babcock, so my assumption would be that Glides, working off the rating, would not recognize any of the dispatched trips as GTFS trips (because they all go to Boston College) and would publish all ADDED trips. Am I thinking about this wrong?

Reply:
NB for the Transit Data folks that, iff it's decided that there isn't value in trying to match to scheduled disrupted trips in this case (or even if there is!), we should take the opportunity to consider making use of the `DELETED` ScheduleRelationship (https://gtfs.org/realtime/reference/#enum-schedulerelationship_1).